app/src/main/java/com/celzero/bravedns/service/ConnectionMonitor.kt

```diff
@@ -0,0 +1,264 @@
+/*
+ * Copyright 2021 RethinkDNS and its authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.celzero.bravedns.service
+
+import android.content.Context
+import android.net.*
+import android.os.Handler
+import android.os.HandlerThread
+import android.os.Looper
+import android.os.Message
+import android.util.Log
+import com.celzero.bravedns.ui.HomeScreenActivity.GlobalVariable.DEBUG
+import com.celzero.bravedns.util.Constants.Companion.LOG_TAG
+import com.google.common.collect.Sets
+import org.koin.core.component.KoinComponent
+import org.koin.core.component.inject
+
+
+class ConnectionMonitor(context: Context, val networkListener: NetworkListener) : ConnectivityManager.NetworkCallback(), KoinComponent {
+    private val networkRequest: NetworkRequest = NetworkRequest.Builder().addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET).build()
+
+    @Volatile private var handlerThread: HandlerThread? = null
```

Comment:
Android handler thread internally operates on a looper. Defined as @Volatile so that the looper in the Handler thread so that the value loaded from memory instead of Thread's local cache. Referred many code samples, mostly they are made as volatile else the handler's looper is made as volatile.

Reply:
Added reference in the comment.